src/main/scala/com/twitter/finagle/postgres/codec/PgCodec.scala

```diff
@@ -120,63 +115,46 @@ class AuthenticationProxy(
 
-  override def messageReceived(ctx: ChannelHandlerContext, e: MessageEvent) {
-    val message = e.getMessage
-
-    message match {
-      case packet: Packet =>
-        parser.parse(packet) match {
-          case Some(backendMessage) =>
-            Channels.fireMessageReceived(ctx, backendMessage)
-          case None =>
-            logger.warning("Cannot parse the packet. Disconnecting...")
-            Channels.disconnect(ctx.getChannel)
-        }
-
-      case _ =>
-        logger.warning("Only packet is supported...")
-        Channels.disconnect(ctx.getChannel)
-    }
+  override def decode(ctx: ChannelHandlerContext, msg: Packet, out: util.List[AnyRef]): Unit = parser.parse(msg) match {
+    case Some(backendMessage) =>
+      out.add(backendMessage)
+    case None => logger.warning("Cannot parse the packet. Disconnecting...")
   }
 }
 
 /*
  * Decodes a byte stream into a Packet.
  */
-class PacketDecoder(@volatile var inSslNegotation: Boolean) extends FrameDecoder {
+class PacketDecoder(@volatile var inSslNegotation: Boolean) extends ByteToMessageDecoder {
   private val logger = Logger(getClass.getName)
 
-  def decode(ctx: ChannelHandlerContext, channel: Channel, buffer: ChannelBuffer): AnyRef = {
-    if (inSslNegotation && buffer.readableBytes() >= 1) {
-      val SslCode: Char = buffer.readByte().asInstanceOf[Char]
+  // TODO: SSL negoatiation should be moved upstream; then this could be a LengthFieldBasedFrameDecoder.
+  override def decode(ctx: ChannelHandlerContext, buffer: ByteBuf, out: util.List[AnyRef]): Unit = if (inSslNegotation && buffer.readableBytes() >= 1) {
 
-      logger.ifDebug("Got ssl negotiation char packet: %s".format(SslCode))
+    val SslCode: Char = buffer.readByte().asInstanceOf[Char]
+    logger.ifDebug("Got ssl negotiation char packet: %s".format(SslCode))
+    inSslNegotation = false
 
-      inSslNegotation = false
+    out.add(new Packet(Some(SslCode), 1, null, true))
 
-      new Packet(Some(SslCode), 1, null, true)
-    } else if (buffer.readableBytes() < 5) {
-      null
-    } else {
-      buffer.markReaderIndex()
-      val code: Char = buffer.readByte().asInstanceOf[Char]
+  } else if (buffer.readableBytes() >= 5) {
+    buffer.markReaderIndex()
+    val code: Char = buffer.readByte().asInstanceOf[Char]
 
-      val totalLength = buffer.readInt()
-      val length = totalLength - 4
+    val totalLength = buffer.readInt()
+    val length = totalLength - 4
 
-      if (buffer.readableBytes() < length) {
-        buffer.resetReaderIndex()
-        return null
-      }
-
-      val packet = new Packet(Some(code), totalLength, buffer.readSlice(length))
-
-      packet
+    if (buffer.readableBytes() < length) {
+      buffer.resetReaderIndex()
+    } else {
+      out.add(new Packet(Some(code), totalLength, buffer.readBytes(length)))
+      // TODO (jeremyrsmith): can we slice here? I tried using readSlice (illegal ref counts everywhere) and readRetainedSlice (severe leak warning).
```

Comment:
Can you try `retainedSlice`?

Reply:
I did try `readRetainedSlice`, which AFAICT is equivalent to `retainedSlice` from the current position.

Reply:
I'm a little surprised that there isn't leak warnings with this as well since _I think_ you're using a pooled allocator (that is the default that Finagle uses) and `.readBytes(length)` should be reusing the parent allocator to allocate a new, likely pooled, `ByteBuf`.

Reply:
TBH I'm not sure of the utility of `Packet`; it predates me. Looking more closely, I think you're right - here we're creating a retained buffer from the input, and because it's coming from input it *won't* get delivered to Netty. I am not sure why using a slice of the input buffer causes `IllegalReferenceCount`s downstream, but it's probably due to asynchrony.

I'll track down where the packet's `content` is actually consumed, and see if adding a `release` there causes any trouble.

Reply:
You need to take a `readRetainedSlice`: the Netty4 `ByteToMessageCodec.decode` method doesn't get ownership of the incoming `ByteBuf` so it's the implementer's responsibility to retain any slices/copy any data they wish to hold onto. You're making a copy right now, which is fine (albeit less efficient), but is just as much a leak as retaining a slice of the parent buffer.